File: scrapers/etoland.py
import asyncio
import re

import httpx

from scrapers.base import BaseScraper, Post
from utils.http_client import fetch_html

_NUM_RE = re.compile(r"(\d[\d,]*)")
_BO_RE = re.compile(r'g4_bo_table\s*=\s*"([^"]+)"')
_WR_RE = re.compile(r'bo_table=\w+&wr_id=(\d+)')


class EtolandScraper(BaseScraper):
    community = "etoland"
    community_name = "이토랜드"
    base_url = "https://www.etoland.co.kr"
    encoding = "euc-kr"

    async def _resolve_real_url(self, client: httpx.AsyncClient, hit_url: str) -> str | None:
        """hit.php?bn_id= 래퍼 URL → board.php 실제 URL 변환"""
        try:
            html = await fetch_html(client, hit_url, encoding=self.encoding, referer=self.base_url)
            bo = _BO_RE.search(html)
            wr = _WR_RE.search(html)
            if bo and wr:
                return f"{self.base_url}/bbs/board.php?bo_table={bo.group(1)}&wr_id={wr.group(1)}"
        except Exception:
            pass
        return None

    async def fetch_best_posts(self, client: httpx.AsyncClient) -> list[Post]:
        soup = await self._fetch_html(client, f"{self.base_url}/bbs/hit.php")
        # 1단계: 히트 목록에서 기본 정보 수집
        raw_posts = []
        seen_bn = set()

        for li in soup.find_all("li", class_=lambda c: c and "hit_item" in c):
            cls = " ".join(li.get("class", []))
            if "ad_list" in cls:
                continue

            a_tag = li.find("a", class_="content_link")
            if not a_tag:
                continue
            subj = li.find("p", class_="subject")
            if not subj:
                continue
            title = subj.get_text(strip=True)
            href = a_tag.get("href", "")
            if not title or not href:
                continue

            # 추천수
            votes = 0
            good_span = li.find("span", class_="good")
            if good_span:
                m = _NUM_RE.search(good_span.get_text(strip=True))
                if m:
                    votes = int(m.group(1).replace(",", ""))

            # 조회수
            views = 0
            hit_span = li.find("span", class_="hit")
            if hit_span:
                m = _NUM_RE.search(hit_span.get_text(strip=True))
                if m:
                    views = int(m.group(1).replace(",", ""))

            # 댓글수
            comments = 0
            cmt_span = li.find("span", class_="comment_cnt")
            if cmt_span:
                m = _NUM_RE.search(cmt_span.get_text(strip=True))
                if m:
                    comments = int(m.group(1).replace(",", ""))

            # hit.php?bn_id= 형태 → 절대 URL
            if href.startswith("http"):
                hit_url = href
            elif href.startswith("/"):
                hit_url = f"{self.base_url}{href}"
            else:
                hit_url = f"{self.base_url}/bbs/{href}"

            if hit_url in seen_bn:
                continue
            seen_bn.add(hit_url)
            raw_posts.append((title, hit_url, votes, views, comments))

        # 2단계: hit.php?bn_id= URL → board.php 실제 URL 병렬 변환
        tasks = [self._resolve_real_url(client, hit_url) for _, hit_url, *_ in raw_posts]
        resolved = await asyncio.gather(*tasks)

        posts = []
        seen_url = set()
        for (title, hit_url, votes, views, comments), real_url in zip(raw_posts, resolved):
            url = real_url or hit_url  # 변환 실패 시 원본 유지
            if url not in seen_url:
                seen_url.add(url)
                posts.append(self._make_post(title, url, votes=votes, views=views,
                                             comments=comments))

        return posts

File: internal/scraper/etoland.go
package scraper

import (
	"fmt"
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"sync"

	"github.com/PuerkitoBio/goquery"
)

var (
	etolandNumRe  = regexp.MustCompile(`(\d[\d,]*)`)
	etolandGoodRe = regexp.MustCompile(`mw\.good\.php\?bo_table=(\w+)&wr_id=(\d+)`)
)

// EtolandScraper scrapes 이토랜드 hit posts (EUC-KR, URL resolution).
type EtolandScraper struct {
	baseScraper
}

func NewEtolandScraper() *EtolandScraper {
	return &EtolandScraper{
		baseScraper: baseScraper{
			community:     "etoland",
			communityName: "이토랜드",
			baseURL:       "https://www.etoland.co.kr",
			encoding:      "euc-kr",
		},
	}
}

func (s *EtolandScraper) Name() string { return s.communityName }

// resolveRealURL converts hit.php?bn_id= URL to board.php real URL.
func (s *EtolandScraper) resolveRealURL(client *http.Client, hitURL string) string {
	html, err := fetchHTML(client, hitURL, s.encoding, s.baseURL)
	if err != nil {
		return ""
	}
	m := etolandGoodRe.FindStringSubmatch(html)
	if m != nil {
		return fmt.Sprintf("%s/bbs/board.php?bo_table=%s&wr_id=%s", s.baseURL, m[1], m[2])
	}
	return ""
}

func (s *EtolandScraper) FetchBestPosts(client *http.Client) ([]Post, error) {
	doc, err := fetchDocument(client, fmt.Sprintf("%s/bbs/hit.php", s.baseURL), s.encoding, s.baseURL)
	if err != nil {
		return nil, err
	}

	type rawPost struct {
		title    string
		hitURL   string
		votes    int
		views    int
		comments int
	}
	var rawPosts []rawPost
	seenBn := make(map[string]bool)

	doc.Find("li").Each(func(_ int, li *goquery.Selection) {
		classes, _ := li.Attr("class")
		if !strings.Contains(classes, "hit_item") {
			return
		}
		if strings.Contains(classes, "ad_list") {
			return
		}

		aTag := li.Find("a.content_link").First()
		if aTag.Length() == 0 {
			return
		}
		subj := li.Find("p.subject").First()
		if subj.Length() == 0 {
			return
		}
		title := strings.TrimSpace(subj.Text())
		href, exists := aTag.Attr("href")
		if !exists || title == "" || href == "" {
			return
		}

		// Votes
		votes := 0
		goodSpan := li.Find("span.good").First()
		if goodSpan.Length() > 0 {
			m := etolandNumRe.FindStringSubmatch(strings.TrimSpace(goodSpan.Text()))
			if m != nil {
				if v, err := strconv.Atoi(strings.ReplaceAll(m[1], ",", "")); err == nil {
					votes = v
				}
			}
		}

		// Views
		views := 0
		hitSpan := li.Find("span.hit").First()
		if hitSpan.Length() > 0 {
			m := etolandNumRe.FindStringSubmatch(strings.TrimSpace(hitSpan.Text()))
			if m != nil {
				if v, err := strconv.Atoi(strings.ReplaceAll(m[1], ",", "")); err == nil {
					views = v
				}
			}
		}

		// Comments
		comments := 0
		cmtSpan := li.Find("span.comment_cnt").First()
		if cmtSpan.Length() > 0 {
			m := etolandNumRe.FindStringSubmatch(strings.TrimSpace(cmtSpan.Text()))
			if m != nil {
				if v, err := strconv.Atoi(strings.ReplaceAll(m[1], ",", "")); err == nil {
					comments = v
				}
			}
		}

		// Resolve hit URL
		var hitURL string
		if strings.HasPrefix(href, "http") {
			hitURL = href
		} else if strings.HasPrefix(href, "/") {
			hitURL = s.baseURL + href
		} else {
			hitURL = s.baseURL + "/bbs/" + href
		}

		if !seenBn[hitURL] {
			seenBn[hitURL] = true
			rawPosts = append(rawPosts, rawPost{
				title:    title,
				hitURL:   hitURL,
				votes:    votes,
				views:    views,
				comments: comments,
			})
		}
	})

	// Resolve hit.php URLs to board.php URLs concurrently
	resolved := make([]string, len(rawPosts))
	var wg sync.WaitGroup
	for i, rp := range rawPosts {
		wg.Add(1)
		go func(idx int, hitURL string) {
			defer wg.Done()
			resolved[idx] = s.resolveRealURL(client, hitURL)
		}(i, rp.hitURL)
	}
	wg.Wait()

	var posts []Post
	seenURL := make(map[string]bool)
	for i, rp := range rawPosts {
		realURL := resolved[i]
		if realURL == "" {
			continue // resolution failed, skip
		}
		if !seenURL[realURL] {
			seenURL[realURL] = true
			posts = append(posts, s.makePost(rp.title, realURL, rp.votes, rp.views, rp.comments))
		}
	}

	return s.filterPosts(posts), nil
}
